src/sentry/api/endpoints/organization_teams.py

```diff
@@ -142,47 +145,71 @@ def post(self, request: Request, organization, **kwargs) -> Response:
+            raise ValidationError(serializer.errors)
 
-            try:
-                with transaction.atomic():
-                    team = Team.objects.create(
-                        name=result.get("name") or result["slug"],
-                        slug=result.get("slug"),
-                        idp_provisioned=result.get("idp_provisioned", False),
-                        organization=organization,
-                    )
-            except IntegrityError:
-                return Response(
-                    {
-                        "non_field_errors": [CONFLICTING_SLUG_ERROR],
-                        "detail": CONFLICTING_SLUG_ERROR,
-                    },
-                    status=409,
+        result = serializer.validated_data
+        set_team_admin = result.get("set_team_admin")
+
+        if set_team_admin:
+            if not features.has("organizations:team-roles", organization) or not features.has(
+                "organizations:team-project-creation-all", organization
+            ):
+                raise ResourceDoesNotExist(
+                    detail="You do not have permission to join a new team as a team admin"
+                )
+            if not self.should_add_creator_to_team(request):
+                raise ValidationError(
+                    {"detail": "You do not have permission to join a new team as a Team Admin"}
+                )
+        try:
+            # Wrap team creation and member addition in same transaction
+            with transaction.atomic():
+                team = Team.objects.create(
+                    name=result.get("name") or result["slug"],
+                    slug=result.get("slug"),
+                    idp_provisioned=result.get("idp_provisioned", False),
+                    organization=organization,
                 )
-            else:
                 team_created.send_robust(
-                    organization=organization, user=request.user, team=team, sender=self.__class__
+                    organization=organization,
+                    user=request.user,
+                    team=team,
+                    sender=self.__class__,
                 )
-            if self.should_add_creator_to_team(request):
-                try:
+                if self.should_add_creator_to_team(request):
                     member = OrganizationMember.objects.get(
                         user=request.user, organization=organization
                     )
-                except OrganizationMember.DoesNotExist:
-                    pass
-                else:
-                    OrganizationMemberTeam.objects.create(team=team, organizationmember=member)
-
+                    OrganizationMemberTeam.objects.create(
+                        team=team,
+                        organizationmember=member,
+                        role="admin" if set_team_admin else None,
+                    )
+        except (IntegrityError, MaxSnowflakeRetryError):
```

Comment:
I discovered that attempting to create a team with a duplicate slug leads to a `MaxSnowflakeRetryError` instead of `IntegrityError` b/c of the new [Snowflake ID hybrid cloud changes](https://vanguard.getsentry.net/p/cl8kqyvst201410ls6q9medy20). You can see it catching the `IntegrityError` and throwing the `MaxSnowflakeRetryError` in the [code here](https://github.com/getsentry/sentry/blob/cc82a92ae4cf5237f1a5fca294fe1e3e7a7a10cf/src/sentry/utils/snowflake.py#L29)